{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "516417f8-7ef5-460e-a9dd-33d5f568f074",
   "metadata": {},
   "source": [
    "This notebook is used to separate a list of continuous genomes from the group of isolates we are working with. This is done by reading through each of the files we are concerned with, and counting the occurrences of the '>' character, which indicates a new contig, or strand of DNA representing the genome. A continuous genome would have a single contig.\n",
    "The 'matched' list names all of the genomes previously determined to be in both data sources used in this analysis. It may be subject to change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5238fb81-0463-404b-8ca6-ed4316e30966",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7c01f7be-7f13-4d4c-a678-e828663b6f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "REFS = '../data/img_data_cycog6/'\n",
    "matched = ['AS9601', 'EQPAC1', 'GP2', 'LG', 'MED4', 'MIT0601', 'MIT0603', 'MIT0604', 'MIT0701', 'MIT0702', 'MIT0703', 'MIT0801', 'MIT0912', 'MIT0913', 'MIT0915', 'MIT0917', 'MIT0918', 'MIT0919', 'MIT1013', 'MIT1205', 'MIT1214', 'MIT1223', 'MIT1300', 'MIT1303', 'MIT1304', 'MIT1306', 'MIT1307', 'MIT1312', 'MIT1313', 'MIT1314', 'MIT1318', 'MIT1320', 'MIT1323', 'MIT1327', 'MIT1341', 'MIT1342', 'MIT1418', 'MIT9107', 'MIT9116', 'MIT9123', 'MIT9202', 'MIT9211', 'MIT9215', 'MIT9301', 'MIT9302', 'MIT9303', 'MIT9311', 'MIT9312', 'MIT9313', 'MIT9314', 'MIT9321', 'MIT9322', 'MIT9401', 'MIT9515', 'NATL1A', 'NATL2A', 'PAC1', 'SB', 'SS120', 'SS2', 'SS35', 'SS51', 'SS52']\n",
    "datapath99 = ('../data/cycogs/cycogsgenomes.tsv') # basic information on each genome\n",
    "genomes_df = pd.read_csv(datapath99, sep='\\t')\n",
    "matched_IMGs = genomes_df[genomes_df.IID.isin(matched)].IMG_ID.tolist()\n",
    "matched_IMGs_str = [str(i) for i in matched_IMGs]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "54861baf-e39b-4ba4-9196-c3ebbe0cd717",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2681813566</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2606217692</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2681813568</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2623620348</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2623620984</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>2623620959</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2606217419</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>2606217685</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>2681813571</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>2606217682</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome  Count\n",
       "0   2681813566     45\n",
       "1   2606217692     13\n",
       "2   2681813568      1\n",
       "3   2623620348      1\n",
       "4   2623620984      6\n",
       "..         ...    ...\n",
       "58  2623620959      1\n",
       "59  2606217419     20\n",
       "60  2606217685     14\n",
       "61  2681813571      8\n",
       "62  2606217682     61\n",
       "\n",
       "[63 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make a dataframe matching genome ID to contig count for the matched isolates\n",
    "data = []\n",
    "\n",
    "# Iterate over the directories within REFS\n",
    "for genome in os.listdir(REFS):\n",
    "    genome_path = os.path.join(REFS, genome)\n",
    "    \n",
    "    # Skip non-directory entries if any\n",
    "    if not os.path.isdir(genome_path):\n",
    "        continue\n",
    "    \n",
    "    if genome in matched_IMGs_str:\n",
    "        # Iterate over files within each genome directory\n",
    "        for file_name in os.listdir(genome_path):\n",
    "            # Check if the file has a .fna extension and does not end with 'genes.fna' or 'intergenic.fna'\n",
    "            if file_name.endswith('.fna') and not (file_name.endswith('genes.fna') or file_name.endswith('intergenic.fna')):\n",
    "                file_path = os.path.join(genome_path, file_name)\n",
    "                # Open and read the .fna file\n",
    "                with open(file_path, 'r') as fna_file:\n",
    "                    content = fna_file.read()\n",
    "                    # Count the number of occurrences of the '>' character\n",
    "                    count_greater_than = content.count('>')\n",
    "                    # Append the result to the data list\n",
    "                    data.append([genome, count_greater_than])\n",
    "\n",
    "# Create a DataFrame from the data list\n",
    "df = pd.DataFrame(data, columns=['Genome', 'Count'])\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56a98b98-f04a-4eec-9292-0772da192e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# converts genome ID to integers, necessary for further analysis\n",
    "imgs = [eval(i) for i in df.Genome.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "07ee6194-762f-402c-ae6e-a31c6fe6f370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IID</th>\n",
       "      <th>IMG_ID</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIT1205</td>\n",
       "      <td>2681813566</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT9107</td>\n",
       "      <td>2606217692</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIT1223</td>\n",
       "      <td>2681813568</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NATL1A</td>\n",
       "      <td>2623620348</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIT9202</td>\n",
       "      <td>2623620984</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>2623620959</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PAC1</td>\n",
       "      <td>2606217419</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>LG</td>\n",
       "      <td>2606217685</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MIT1304</td>\n",
       "      <td>2681813571</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MIT0703</td>\n",
       "      <td>2606217682</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        IID      IMG_ID  Count\n",
       "0   MIT1205  2681813566     45\n",
       "1   MIT9107  2606217692     13\n",
       "2   MIT1223  2681813568      1\n",
       "3    NATL1A  2623620348      1\n",
       "4   MIT9202  2623620984      6\n",
       "..      ...         ...    ...\n",
       "58   AS9601  2623620959      1\n",
       "59     PAC1  2606217419     20\n",
       "60       LG  2606217685     14\n",
       "61  MIT1304  2681813571      8\n",
       "62  MIT0703  2606217682     61\n",
       "\n",
       "[63 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run this block twice\n",
    "# adds the genome name to the dataframe\n",
    "imgs = df.Genome.tolist()\n",
    "data = []\n",
    "# Iterate over genomes_df rows\n",
    "for index, value in genomes_df.iterrows():\n",
    "    if value['IMG_ID'] in imgs:\n",
    "        data.append([value['IID'], value['IMG_ID']])\n",
    "\n",
    "# Create a DataFrame from the data list\n",
    "result_df = pd.DataFrame(data, columns=['IID', 'IMG_ID'])\n",
    "df['Genome'] = df['Genome'].astype(int)\n",
    "\n",
    "# Merge result_df with df\n",
    "merged_df = df.merge(result_df, left_on='Genome', right_on='IMG_ID')\n",
    "\n",
    "merged_df = merged_df.drop(['Genome'], axis=1)[['IID', 'IMG_ID', 'Count']]\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "feb19264-98ad-4376-ab02-bbb0c480ebfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MIT1223', 'NATL1A', 'MED4', 'MIT1214', 'MIT0913', 'MIT0915', 'MIT9215', 'MIT0801', 'MIT9312', 'MIT1314', 'MIT1341', 'MIT0917', 'MIT1307', 'MIT0604', 'MIT9303', 'MIT9515', 'MIT1013', 'MIT0919', 'MIT0918', 'MIT9313', 'NATL2A', 'MIT9211', 'MIT1300', 'MIT0912', 'MIT9301', 'SS120', 'AS9601']\n"
     ]
    }
   ],
   "source": [
    "# list to be copied to ortholog_census.ipynb to isolate continuous genomes\n",
    "single_contig_df = merged_df[merged_df.Count == 1]\n",
    "single_contigs = single_contig_df.IID.tolist()\n",
    "print (single_contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af5a2d5e-5bac-4bce-8d25-a67a34b8d740",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
