{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "516417f8-7ef5-460e-a9dd-33d5f568f074",
   "metadata": {},
   "source": [
    "# Data Inputs\n",
    "(adaped from `continuous_filtering.ipynb`)\n",
    "\n",
    "The purpose of this notebook is to assemble all of the input files needed to run our analysis of genomic islands (scripts in this directory). The following files were downloaded from the links listed to the locations indicated:\n",
    "\n",
    "1. `Data_S1-Pro-Clusters.xlsx` from Supplementary Materials of [Blaskowski et al., 2024](https://zenodo.org/records/12210994)\n",
    "    - Downloaded to `data/clusters/`\n",
    "1. `Data_S7-genome-metadata.csv` from Supplementary Materials of [Blaskowski et al., 2024](https://zenodo.org/records/12210994)\n",
    "    - Downloaded to `data/cycogs/`\n",
    "1. `Data_S8-img_data_cycog6.tar.gz` from Supplementary Materials of [Blaskowski et al., 2024](https://zenodo.org/records/12210994)\n",
    "    - Downloaded to this directory (`data/`) and extracted with `tar -xzvf Data_S8-img_data_cycog6.tar.gz`\n",
    "    - Extracted directory is named `data/img_data_cycog6/`\n",
    "1. `cycogs.tsv` from Supplementary Materials of [Berube et al., 2018](https://doi.org/10.6084/m9.figshare.c.4037048.v1)\n",
    "    - Downloaded to `data/cycogs/`\n",
    "1. `mmc1.xlsx` and `mmc2.xlsx` from Supplementary Materials of [Hackl et al., 2023](https://doi.org/10.1016/j.cell.2022.12.006)\n",
    "    - Downloaded to `data/hackl-2023/`\n",
    "\n",
    "The outputs of this notebook are the following files:\n",
    "\n",
    "1. `genome-metadata.csv`\n",
    "    - A file of all of the _Prochlorococcus_ genomes used to make CyCOGv6, with additional information added in, including:\n",
    "        - Whether or not a genome was analyzed for predicted genomic islands by Hackl et al.\n",
    "        - The number of contigs per genome in the sequences used to make CyCOGv6 (`Data_S8-img_data_cycog6.tar.gz`)\n",
    "1. `ortholog-metadata.csv`\n",
    "    - A file of all of the genes included in CyCOGv6 that includes the following fields:\n",
    "       - `MappingName`, `CyCOGID`, `GenomeName`, `GeneID`, `GeneID`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5238fb81-0463-404b-8ca6-ed4316e30966",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "464b30dc-27bc-4100-85ce-a5751409b5c2",
   "metadata": {},
   "source": [
    "## Genome Metadata\n",
    "\n",
    "This code block builds the `genome-metadata.csv` file from the `cycogs-genomes.tsv` and other input files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d805d3f9-13c3-49e3-af58-59630900ccf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Type</th>\n",
       "      <th>Group</th>\n",
       "      <th>CladeCyCOG</th>\n",
       "      <th>Virocell</th>\n",
       "      <th>CompletenessCyCOG</th>\n",
       "      <th>CyCOGGenome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884698</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884700</td>\n",
       "      <td>AG-316-N23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884701</td>\n",
       "      <td>AG-316-P23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>12.07</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884699</td>\n",
       "      <td>AG-316-L21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>10.34</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884642</td>\n",
       "      <td>AG-316-A05</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>6.90</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>2681813574</td>\n",
       "      <td>MIT1341</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>100.00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>2681813570</td>\n",
       "      <td>MIT1300</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.86</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>2681813572</td>\n",
       "      <td>MIT1307</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.18</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>2667527276</td>\n",
       "      <td>AG-402-N21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVIII</td>\n",
       "      <td>False</td>\n",
       "      <td>79.80</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>2681813568</td>\n",
       "      <td>MIT1223</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVIII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.18</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>604 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID  GenomeName     Type            Group CladeCyCOG  Virocell  \\\n",
       "0    2716884698  AG-316-L16      SAG  Prochlorococcus     AMZ-II     False   \n",
       "1    2716884700  AG-316-N23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "2    2716884701  AG-316-P23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "3    2716884699  AG-316-L21      SAG  Prochlorococcus     AMZ-II     False   \n",
       "4    2716884642  AG-316-A05      SAG  Prochlorococcus     AMZ-II     False   \n",
       "..          ...         ...      ...              ...        ...       ...   \n",
       "599  2681813574     MIT1341  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "600  2681813570     MIT1300  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "601  2681813572     MIT1307  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "602  2667527276  AG-402-N21      SAG  Prochlorococcus     LLVIII     False   \n",
       "603  2681813568     MIT1223  ISOLATE  Prochlorococcus     LLVIII     False   \n",
       "\n",
       "     CompletenessCyCOG  CyCOGGenome  \n",
       "0                20.69         True  \n",
       "1                20.69         True  \n",
       "2                12.07         True  \n",
       "3                10.34         True  \n",
       "4                 6.90         True  \n",
       "..                 ...          ...  \n",
       "599             100.00         True  \n",
       "600              99.86         True  \n",
       "601              99.18         True  \n",
       "602              79.80         True  \n",
       "603              99.18         True  \n",
       "\n",
       "[604 rows x 8 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in genome metadata from Blaskowski et al.\n",
    "\n",
    "cycog_genome_df = pd.read_csv('cycogs/Data_S7-genome-metadata.csv')\n",
    "cycog_genome_df = cycog_genome_df.rename(columns={'Clade': 'CladeCyCOG', 'Completeness': 'CompletenessCyCOG'})\n",
    "cycog_genome_df = cycog_genome_df[cycog_genome_df.Group == 'Prochlorococcus']\n",
    "cycog_genome_df['CyCOGGenome'] = True\n",
    "\n",
    "cycog_genome_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2af08bb7-2bf9-4c92-af29-4a7dcc26bcc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Type</th>\n",
       "      <th>Group</th>\n",
       "      <th>CladeCyCOG</th>\n",
       "      <th>Virocell</th>\n",
       "      <th>CompletenessCyCOG</th>\n",
       "      <th>CyCOGGenome</th>\n",
       "      <th>NContigsCyCOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884698</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884700</td>\n",
       "      <td>AG-316-N23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884701</td>\n",
       "      <td>AG-316-P23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>12.07</td>\n",
       "      <td>True</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884699</td>\n",
       "      <td>AG-316-L21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>10.34</td>\n",
       "      <td>True</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884642</td>\n",
       "      <td>AG-316-A05</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>6.90</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>2681813574</td>\n",
       "      <td>MIT1341</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>100.00</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>2681813570</td>\n",
       "      <td>MIT1300</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.86</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>2681813572</td>\n",
       "      <td>MIT1307</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.18</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>2667527276</td>\n",
       "      <td>AG-402-N21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVIII</td>\n",
       "      <td>False</td>\n",
       "      <td>79.80</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>2681813568</td>\n",
       "      <td>MIT1223</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLVIII</td>\n",
       "      <td>False</td>\n",
       "      <td>99.18</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>604 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID  GenomeName     Type            Group CladeCyCOG  Virocell  \\\n",
       "0    2716884698  AG-316-L16      SAG  Prochlorococcus     AMZ-II     False   \n",
       "1    2716884700  AG-316-N23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "2    2716884701  AG-316-P23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "3    2716884699  AG-316-L21      SAG  Prochlorococcus     AMZ-II     False   \n",
       "4    2716884642  AG-316-A05      SAG  Prochlorococcus     AMZ-II     False   \n",
       "..          ...         ...      ...              ...        ...       ...   \n",
       "599  2681813574     MIT1341  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "600  2681813570     MIT1300  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "601  2681813572     MIT1307  ISOLATE  Prochlorococcus      LLVII     False   \n",
       "602  2667527276  AG-402-N21      SAG  Prochlorococcus     LLVIII     False   \n",
       "603  2681813568     MIT1223  ISOLATE  Prochlorococcus     LLVIII     False   \n",
       "\n",
       "     CompletenessCyCOG  CyCOGGenome  NContigsCyCOG  \n",
       "0                20.69         True             64  \n",
       "1                20.69         True             41  \n",
       "2                12.07         True             31  \n",
       "3                10.34         True             34  \n",
       "4                 6.90         True             20  \n",
       "..                 ...          ...            ...  \n",
       "599             100.00         True              1  \n",
       "600              99.86         True              1  \n",
       "601              99.18         True              1  \n",
       "602              79.80         True             47  \n",
       "603              99.18         True              1  \n",
       "\n",
       "[604 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add in information on how many contigs each CyCOG genome has\n",
    "\n",
    "genome_dir = 'img_data_cycog6/'\n",
    "\n",
    "# make a dataframe matching genome ID to contig count for the matched isolates\n",
    "data = []\n",
    "\n",
    "# Iterate over the directories within REFS\n",
    "for genome in os.listdir(genome_dir):\n",
    "    # print(genome)\n",
    "    genome_path = os.path.join(genome_dir, genome)\n",
    "    \n",
    "    # Skip non-directory entries if any\n",
    "    if not os.path.isdir(genome_path):\n",
    "        continue\n",
    "    \n",
    "    if int(genome) in cycog_genome_df['GenomeID'].tolist():\n",
    "        # Build fasta filename of genome\n",
    "        file_path = os.path.join(genome_path, f'{genome}.fna')\n",
    "        # Open and read the .fna file\n",
    "        with open(file_path, 'r') as fna_file:\n",
    "            content = fna_file.read()\n",
    "            # Count the number of occurrences of the '>' character\n",
    "            count_greater_than = content.count('>')\n",
    "            # Append the result to the data list\n",
    "            data.append((int(genome), count_greater_than))\n",
    "\n",
    "# Add count data to dataframe\n",
    "cycog_genome_df['NContigsCyCOG'] = cycog_genome_df['GenomeID'].map(dict(data))\n",
    "cycog_genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "19ffa188-856d-4eff-8f0a-3da88eec7e6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>name</th>\n",
       "      <th>CladeHackl</th>\n",
       "      <th>notes</th>\n",
       "      <th>Type</th>\n",
       "      <th>depth</th>\n",
       "      <th>lat</th>\n",
       "      <th>long</th>\n",
       "      <th>assembly_genome_size</th>\n",
       "      <th>assembly_ambiguous_bases</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonomy_strain_ncbi_taxid</th>\n",
       "      <th>data_assembly_file_origin</th>\n",
       "      <th>data_ncbi_bioproject_accession</th>\n",
       "      <th>data_ncbi_biosample_accession</th>\n",
       "      <th>data_ncbi_genbank_accession</th>\n",
       "      <th>data_ncbi_refseq_accession</th>\n",
       "      <th>data_ncbi_wgs_accession</th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>data_assembly_ftp</th>\n",
       "      <th>GenomicIslands</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-311-I09</td>\n",
       "      <td>Uncultured Prochlorococcus sp. AG-311-I09</td>\n",
       "      <td>HLI</td>\n",
       "      <td>Simons sequencing project</td>\n",
       "      <td>SAG</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>697970</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bigelow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.716885e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-311-J05</td>\n",
       "      <td>Uncultured Prochlorococcus sp. AG-311-J05</td>\n",
       "      <td>HLI</td>\n",
       "      <td>Simons sequencing project</td>\n",
       "      <td>SAG</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>623148</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bigelow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.716885e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-311-M23</td>\n",
       "      <td>Uncultured Prochlorococcus sp. AG-311-M23</td>\n",
       "      <td>HLI</td>\n",
       "      <td>Simons sequencing project</td>\n",
       "      <td>SAG</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>651213</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bigelow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.716885e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-321-D23</td>\n",
       "      <td>Uncultured Prochlorococcus sp. AG-321-D23</td>\n",
       "      <td>HLI</td>\n",
       "      <td>Simons sequencing project</td>\n",
       "      <td>SAG</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>1201571</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bigelow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.716885e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-321-E21</td>\n",
       "      <td>Uncultured Prochlorococcus sp. AG-321-E21</td>\n",
       "      <td>HLI</td>\n",
       "      <td>Simons sequencing project</td>\n",
       "      <td>SAG</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>1314750</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bigelow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.716885e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>MIT1307</td>\n",
       "      <td>Prochlorococcus sp. MIT1307</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>150.0</td>\n",
       "      <td>22.00</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>2032419</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chisholm Lab</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>MIT1341</td>\n",
       "      <td>Prochlorococcus sp. MIT1341</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolate</td>\n",
       "      <td>150.0</td>\n",
       "      <td>22.00</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>1937096</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chisholm Lab</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>UBA1269</td>\n",
       "      <td>Prochlorococcus sp. UBA1269</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>Tara; location unclear; depth unclear</td>\n",
       "      <td>MAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1606016</td>\n",
       "      <td>71071</td>\n",
       "      <td>...</td>\n",
       "      <td>1947242.0</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>PRJNA348753</td>\n",
       "      <td>SAMN06451922</td>\n",
       "      <td>GCA_002308455.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBVH00000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>UBA1273</td>\n",
       "      <td>Prochlorococcus sp. UBA1273</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>Tara; location unclear; depth unclear</td>\n",
       "      <td>MAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1658888</td>\n",
       "      <td>58506</td>\n",
       "      <td>...</td>\n",
       "      <td>1947243.0</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>PRJNA348753</td>\n",
       "      <td>SAMN06454572</td>\n",
       "      <td>GCA_002308835.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBVD00000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>UBA1528</td>\n",
       "      <td>Prochlorococcus sp. UBA1528</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>Tara; location unclear; depth unclear</td>\n",
       "      <td>MAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2059346</td>\n",
       "      <td>64393</td>\n",
       "      <td>...</td>\n",
       "      <td>1947244.0</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>PRJNA348753</td>\n",
       "      <td>SAMN06457348</td>\n",
       "      <td>GCA_002323815.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DCQC00000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>623 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GenomeName                                       name CladeHackl  \\\n",
       "0    AG-311-I09  Uncultured Prochlorococcus sp. AG-311-I09        HLI   \n",
       "1    AG-311-J05  Uncultured Prochlorococcus sp. AG-311-J05        HLI   \n",
       "2    AG-311-M23  Uncultured Prochlorococcus sp. AG-311-M23        HLI   \n",
       "3    AG-321-D23  Uncultured Prochlorococcus sp. AG-321-D23        HLI   \n",
       "4    AG-321-E21  Uncultured Prochlorococcus sp. AG-321-E21        HLI   \n",
       "..          ...                                        ...        ...   \n",
       "618     MIT1307                Prochlorococcus sp. MIT1307      LLVII   \n",
       "619     MIT1341                Prochlorococcus sp. MIT1341      LLVII   \n",
       "620     UBA1269                Prochlorococcus sp. UBA1269      LLVII   \n",
       "621     UBA1273                Prochlorococcus sp. UBA1273      LLVII   \n",
       "622     UBA1528                Prochlorococcus sp. UBA1528      LLVII   \n",
       "\n",
       "                                     notes     Type  depth    lat    long  \\\n",
       "0                Simons sequencing project      SAG   20.0 -20.08  -70.80   \n",
       "1                Simons sequencing project      SAG   20.0 -20.08  -70.80   \n",
       "2                Simons sequencing project      SAG   20.0 -20.08  -70.80   \n",
       "3                Simons sequencing project      SAG   14.0 -23.46  -88.77   \n",
       "4                Simons sequencing project      SAG   14.0 -23.46  -88.77   \n",
       "..                                     ...      ...    ...    ...     ...   \n",
       "618                                    NaN  Isolate  150.0  22.00 -158.00   \n",
       "619                                    NaN  Isolate  150.0  22.00 -158.00   \n",
       "620  Tara; location unclear; depth unclear      MAG    NaN    NaN     NaN   \n",
       "621  Tara; location unclear; depth unclear      MAG    NaN    NaN     NaN   \n",
       "622  Tara; location unclear; depth unclear      MAG    NaN    NaN     NaN   \n",
       "\n",
       "     assembly_genome_size  assembly_ambiguous_bases  ...  \\\n",
       "0                  697970                         0  ...   \n",
       "1                  623148                         0  ...   \n",
       "2                  651213                         0  ...   \n",
       "3                 1201571                         0  ...   \n",
       "4                 1314750                         0  ...   \n",
       "..                    ...                       ...  ...   \n",
       "618               2032419                         0  ...   \n",
       "619               1937096                         0  ...   \n",
       "620               1606016                     71071  ...   \n",
       "621               1658888                     58506  ...   \n",
       "622               2059346                     64393  ...   \n",
       "\n",
       "     taxonomy_strain_ncbi_taxid  data_assembly_file_origin  \\\n",
       "0                           NaN                    Bigelow   \n",
       "1                           NaN                    Bigelow   \n",
       "2                           NaN                    Bigelow   \n",
       "3                           NaN                    Bigelow   \n",
       "4                           NaN                    Bigelow   \n",
       "..                          ...                        ...   \n",
       "618                         NaN               Chisholm Lab   \n",
       "619                         NaN               Chisholm Lab   \n",
       "620                   1947242.0                       NCBI   \n",
       "621                   1947243.0                       NCBI   \n",
       "622                   1947244.0                       NCBI   \n",
       "\n",
       "     data_ncbi_bioproject_accession  data_ncbi_biosample_accession  \\\n",
       "0                               NaN                            NaN   \n",
       "1                               NaN                            NaN   \n",
       "2                               NaN                            NaN   \n",
       "3                               NaN                            NaN   \n",
       "4                               NaN                            NaN   \n",
       "..                              ...                            ...   \n",
       "618                             NaN                            NaN   \n",
       "619                             NaN                            NaN   \n",
       "620                     PRJNA348753                   SAMN06451922   \n",
       "621                     PRJNA348753                   SAMN06454572   \n",
       "622                     PRJNA348753                   SAMN06457348   \n",
       "\n",
       "     data_ncbi_genbank_accession  data_ncbi_refseq_accession  \\\n",
       "0                            NaN                         NaN   \n",
       "1                            NaN                         NaN   \n",
       "2                            NaN                         NaN   \n",
       "3                            NaN                         NaN   \n",
       "4                            NaN                         NaN   \n",
       "..                           ...                         ...   \n",
       "618                          NaN                         NaN   \n",
       "619                          NaN                         NaN   \n",
       "620              GCA_002308455.1                         NaN   \n",
       "621              GCA_002308835.1                         NaN   \n",
       "622              GCA_002323815.1                         NaN   \n",
       "\n",
       "     data_ncbi_wgs_accession      GenomeID  \\\n",
       "0                        NaN  2.716885e+09   \n",
       "1                        NaN  2.716885e+09   \n",
       "2                        NaN  2.716885e+09   \n",
       "3                        NaN  2.716885e+09   \n",
       "4                        NaN  2.716885e+09   \n",
       "..                       ...           ...   \n",
       "618                      NaN           NaN   \n",
       "619                      NaN           NaN   \n",
       "620             DBVH00000000           NaN   \n",
       "621             DBVD00000000           NaN   \n",
       "622             DCQC00000000           NaN   \n",
       "\n",
       "                                     data_assembly_ftp  GenomicIslands  \n",
       "0                                                  NaN            True  \n",
       "1                                                  NaN            True  \n",
       "2                                                  NaN            True  \n",
       "3                                                  NaN            True  \n",
       "4                                                  NaN            True  \n",
       "..                                                 ...             ...  \n",
       "618                                                NaN            True  \n",
       "619                                                NaN            True  \n",
       "620  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...            True  \n",
       "621  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...            True  \n",
       "622  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...            True  \n",
       "\n",
       "[623 rows x 41 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in genome data from Hackl et al.\n",
    "\n",
    "gi_locations_df = pd.read_excel('hackl-2023/mmc1.xlsx').rename(columns={\n",
    "    'data_img_genome_id': 'GenomeID', \n",
    "    'genome_id': 'GenomeName', \n",
    "    'clade_bac120': 'CladeHackl', \n",
    "    'sample_type': 'Type', \n",
    "    'assembly_contigs': 'NContigsHackl', \n",
    "    'taxonomy_genus': 'Genus', \n",
    "    'checkm_completeness': 'CompletenessHackl'\n",
    "})\n",
    "gi_locations_df['GenomicIslands'] = True\n",
    "\n",
    "gi_locations_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e8410c1-f9d8-40d0-9ff3-5a302bc657c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Type</th>\n",
       "      <th>CladeCyCOG</th>\n",
       "      <th>CompletenessCyCOG</th>\n",
       "      <th>CyCOGGenome</th>\n",
       "      <th>NContigsCyCOG</th>\n",
       "      <th>CladeHackl</th>\n",
       "      <th>CompletenessHackl</th>\n",
       "      <th>GenomicIslands</th>\n",
       "      <th>NContigsHackl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884698</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>SAG</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "      <td>64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884700</td>\n",
       "      <td>AG-316-N23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>20.69</td>\n",
       "      <td>True</td>\n",
       "      <td>41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884701</td>\n",
       "      <td>AG-316-P23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>12.07</td>\n",
       "      <td>True</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884699</td>\n",
       "      <td>AG-316-L21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>10.34</td>\n",
       "      <td>True</td>\n",
       "      <td>34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884642</td>\n",
       "      <td>AG-316-A05</td>\n",
       "      <td>SAG</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>6.90</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>-1</td>\n",
       "      <td>JGI_02_N20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>30.07</td>\n",
       "      <td>True</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>-1</td>\n",
       "      <td>TMED223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>50.40</td>\n",
       "      <td>True</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>-1</td>\n",
       "      <td>UBA1269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>65.70</td>\n",
       "      <td>True</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>-1</td>\n",
       "      <td>UBA1273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>69.16</td>\n",
       "      <td>True</td>\n",
       "      <td>398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>-1</td>\n",
       "      <td>UBA1528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>79.36</td>\n",
       "      <td>True</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>737 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID  GenomeName Type CladeCyCOG  CompletenessCyCOG  CyCOGGenome  \\\n",
       "0    2716884698  AG-316-L16  SAG     AMZ-II              20.69         True   \n",
       "1    2716884700  AG-316-N23  SAG     AMZ-II              20.69         True   \n",
       "2    2716884701  AG-316-P23  SAG     AMZ-II              12.07         True   \n",
       "3    2716884699  AG-316-L21  SAG     AMZ-II              10.34         True   \n",
       "4    2716884642  AG-316-A05  SAG     AMZ-II               6.90         True   \n",
       "..          ...         ...  ...        ...                ...          ...   \n",
       "732          -1  JGI_02_N20  NaN        NaN                NaN        False   \n",
       "733          -1     TMED223  NaN        NaN                NaN        False   \n",
       "734          -1     UBA1269  NaN        NaN                NaN        False   \n",
       "735          -1     UBA1273  NaN        NaN                NaN        False   \n",
       "736          -1     UBA1528  NaN        NaN                NaN        False   \n",
       "\n",
       "     NContigsCyCOG CladeHackl  CompletenessHackl  GenomicIslands  \\\n",
       "0               64        NaN                NaN           False   \n",
       "1               41        NaN                NaN           False   \n",
       "2               31        NaN                NaN           False   \n",
       "3               34        NaN                NaN           False   \n",
       "4               20        NaN                NaN           False   \n",
       "..             ...        ...                ...             ...   \n",
       "732             -1       LLIV              30.07            True   \n",
       "733             -1       LLIV              50.40            True   \n",
       "734             -1      LLVII              65.70            True   \n",
       "735             -1      LLVII              69.16            True   \n",
       "736             -1      LLVII              79.36            True   \n",
       "\n",
       "     NContigsHackl  \n",
       "0               -1  \n",
       "1               -1  \n",
       "2               -1  \n",
       "3               -1  \n",
       "4               -1  \n",
       "..             ...  \n",
       "732             53  \n",
       "733            121  \n",
       "734            440  \n",
       "735            398  \n",
       "736            374  \n",
       "\n",
       "[737 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge \n",
    "\n",
    "genome_df = pd.merge(\n",
    "    cycog_genome_df[['GenomeID', 'GenomeName', 'Type', 'CladeCyCOG', 'CompletenessCyCOG', 'CyCOGGenome', 'NContigsCyCOG']], \n",
    "    gi_locations_df[['GenomeName',  'CladeHackl', 'CompletenessHackl', 'GenomicIslands', 'NContigsHackl']], \n",
    "    on=['GenomeName'], \n",
    "    how='outer'\n",
    ")\n",
    "genome_df['GenomeID'] = genome_df['GenomeID'].fillna(-1).astype(int)\n",
    "genome_df['NContigsCyCOG'] = genome_df['NContigsCyCOG'].fillna(-1).astype(int)\n",
    "genome_df['NContigsHackl'] = genome_df['NContigsHackl'].fillna(-1).astype(int)\n",
    "genome_df['CyCOGGenome'] = genome_df['CyCOGGenome'].fillna(False).astype(bool)\n",
    "genome_df['GenomicIslands'] = genome_df['GenomicIslands'].fillna(False).astype(bool)\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe96f523-554e-4b2c-aa26-149068c6982a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28 genomes are discordant\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>CladeCyCOG</th>\n",
       "      <th>NContigsCyCOG</th>\n",
       "      <th>CladeHackl</th>\n",
       "      <th>NContigsHackl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>2606217689</td>\n",
       "      <td>EQPAC1</td>\n",
       "      <td>HLI</td>\n",
       "      <td>8</td>\n",
       "      <td>HLI</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>2606217677</td>\n",
       "      <td>SB</td>\n",
       "      <td>HLII</td>\n",
       "      <td>4</td>\n",
       "      <td>HLII</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>2606217680</td>\n",
       "      <td>MIT9311</td>\n",
       "      <td>HLII</td>\n",
       "      <td>17</td>\n",
       "      <td>HLII</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>2606217312</td>\n",
       "      <td>MIT9314</td>\n",
       "      <td>HLII</td>\n",
       "      <td>16</td>\n",
       "      <td>HLII</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>2606217606</td>\n",
       "      <td>GP2</td>\n",
       "      <td>HLII</td>\n",
       "      <td>11</td>\n",
       "      <td>HLII</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>2606217692</td>\n",
       "      <td>MIT9107</td>\n",
       "      <td>HLII</td>\n",
       "      <td>13</td>\n",
       "      <td>HLII</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>2606217690</td>\n",
       "      <td>MIT9116</td>\n",
       "      <td>HLII</td>\n",
       "      <td>22</td>\n",
       "      <td>HLII</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>2606217318</td>\n",
       "      <td>MIT9123</td>\n",
       "      <td>HLII</td>\n",
       "      <td>18</td>\n",
       "      <td>HLII</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>2606217691</td>\n",
       "      <td>MIT9302</td>\n",
       "      <td>HLII</td>\n",
       "      <td>17</td>\n",
       "      <td>HLII</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>2551306550</td>\n",
       "      <td>W11</td>\n",
       "      <td>HLIV</td>\n",
       "      <td>158</td>\n",
       "      <td>HLIII/HLIV</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>2606217419</td>\n",
       "      <td>PAC1</td>\n",
       "      <td>LLI</td>\n",
       "      <td>20</td>\n",
       "      <td>LLI</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>2606217686</td>\n",
       "      <td>MIT0603</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>7</td>\n",
       "      <td>LLII</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2623620960</td>\n",
       "      <td>MIT9211</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>1</td>\n",
       "      <td>LLIII</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>2606217684</td>\n",
       "      <td>MIT0701</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>53</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>2606217681</td>\n",
       "      <td>MIT0702</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>61</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>2681813569</td>\n",
       "      <td>MIT1227</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>91</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>2681812924</td>\n",
       "      <td>MIT1306</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>11</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>2681812925</td>\n",
       "      <td>MIT1312</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>41</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>2681812926</td>\n",
       "      <td>MIT1313</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>26</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>2681812927</td>\n",
       "      <td>MIT1318</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>25</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>2681812949</td>\n",
       "      <td>MIT1327</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>29</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>2681812950</td>\n",
       "      <td>MIT1342</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>22</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>2606217682</td>\n",
       "      <td>MIT0703</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>61</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>2681812923</td>\n",
       "      <td>MIT1303</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>39</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>2681812928</td>\n",
       "      <td>MIT1320</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>20</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>2681812948</td>\n",
       "      <td>MIT1323</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>21</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>2681813566</td>\n",
       "      <td>MIT1205</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>45</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>2681813575</td>\n",
       "      <td>MIT1418</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>73</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID GenomeName  CladeCyCOG  NContigsCyCOG  CladeHackl  \\\n",
       "119  2606217689     EQPAC1         HLI              8         HLI   \n",
       "331  2606217677         SB        HLII              4        HLII   \n",
       "334  2606217680    MIT9311        HLII             17        HLII   \n",
       "336  2606217312    MIT9314        HLII             16        HLII   \n",
       "341  2606217606        GP2        HLII             11        HLII   \n",
       "342  2606217692    MIT9107        HLII             13        HLII   \n",
       "344  2606217690    MIT9116        HLII             22        HLII   \n",
       "345  2606217318    MIT9123        HLII             18        HLII   \n",
       "346  2606217691    MIT9302        HLII             17        HLII   \n",
       "359  2551306550        W11        HLIV            158  HLIII/HLIV   \n",
       "477  2606217419       PAC1         LLI             20         LLI   \n",
       "506  2606217686    MIT0603  LLII.LLIII              7        LLII   \n",
       "515  2623620960    MIT9211  LLII.LLIII              1       LLIII   \n",
       "527  2606217684    MIT0701        LLIV             53        LLIV   \n",
       "528  2606217681    MIT0702        LLIV             61        LLIV   \n",
       "529  2681813569    MIT1227        LLIV             91        LLIV   \n",
       "530  2681812924    MIT1306        LLIV             11        LLIV   \n",
       "531  2681812925    MIT1312        LLIV             41        LLIV   \n",
       "532  2681812926    MIT1313        LLIV             26        LLIV   \n",
       "533  2681812927    MIT1318        LLIV             25        LLIV   \n",
       "534  2681812949    MIT1327        LLIV             29        LLIV   \n",
       "535  2681812950    MIT1342        LLIV             22        LLIV   \n",
       "537  2606217682    MIT0703        LLIV             61        LLIV   \n",
       "538  2681812923    MIT1303        LLIV             39        LLIV   \n",
       "539  2681812928    MIT1320        LLIV             20        LLIV   \n",
       "540  2681812948    MIT1323        LLIV             21        LLIV   \n",
       "541  2681813566    MIT1205        LLIV             45        LLIV   \n",
       "542  2681813575    MIT1418        LLIV             73        LLIV   \n",
       "\n",
       "     NContigsHackl  \n",
       "119              7  \n",
       "331              3  \n",
       "334             15  \n",
       "336             15  \n",
       "341             10  \n",
       "342             12  \n",
       "344             18  \n",
       "345             17  \n",
       "346             15  \n",
       "359             62  \n",
       "477             18  \n",
       "506              5  \n",
       "515              2  \n",
       "527             50  \n",
       "528             55  \n",
       "529             75  \n",
       "530             10  \n",
       "531             33  \n",
       "532             23  \n",
       "533             23  \n",
       "534             23  \n",
       "535             20  \n",
       "537             56  \n",
       "538             35  \n",
       "539             18  \n",
       "540             18  \n",
       "541             59  \n",
       "542            100  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at genomes in which number of CyCOG contigs doesn't match number of Hackl contigs\n",
    "\n",
    "overlap_df = genome_df[genome_df['CyCOGGenome'] & genome_df['GenomicIslands']]\n",
    "discordant_df = overlap_df[overlap_df['NContigsCyCOG'] != overlap_df['NContigsHackl']]\n",
    "print(f'{len(discordant_df)} genomes are discordant')\n",
    "discordant_df[['GenomeID', 'GenomeName', 'CladeCyCOG', 'NContigsCyCOG', 'CladeHackl', 'NContigsHackl']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf2320b1-afeb-47c6-8d5d-e9d8170ebe8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 27 complete isolate genomes that are closed (one contig only)representing 7 distinct clades.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Type</th>\n",
       "      <th>CladeCyCOG</th>\n",
       "      <th>CladeHackl</th>\n",
       "      <th>NContigsCyCOG</th>\n",
       "      <th>NContigsHackl</th>\n",
       "      <th>CompletenessCyCOG</th>\n",
       "      <th>CompletenessHackl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>2623620345</td>\n",
       "      <td>MIT9515</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLI</td>\n",
       "      <td>HLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>2606217259</td>\n",
       "      <td>MED4</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLI</td>\n",
       "      <td>HLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.46</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>2681813573</td>\n",
       "      <td>MIT1314</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>2606217688</td>\n",
       "      <td>MIT0604</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.73</td>\n",
       "      <td>99.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>2606217559</td>\n",
       "      <td>MIT9215</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.73</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>2606217708</td>\n",
       "      <td>MIT9312</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.73</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>2623620959</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.64</td>\n",
       "      <td>99.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>2623620961</td>\n",
       "      <td>MIT9301</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>HLII</td>\n",
       "      <td>HLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.46</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>2623620348</td>\n",
       "      <td>NATL1A</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.73</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>2606217240</td>\n",
       "      <td>NATL2A</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.45</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>2606217560</td>\n",
       "      <td>MIT0801</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>99.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>2681812899</td>\n",
       "      <td>MIT0912</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>99.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>2681812900</td>\n",
       "      <td>MIT0913</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>2681812904</td>\n",
       "      <td>MIT1013</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>2681813567</td>\n",
       "      <td>MIT1214</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>2681812859</td>\n",
       "      <td>MIT0917</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>98.91</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>2681812901</td>\n",
       "      <td>MIT0915</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLI</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>98.64</td>\n",
       "      <td>99.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>2623620733</td>\n",
       "      <td>SS120</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>LLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.86</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>2681812903</td>\n",
       "      <td>MIT0919</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>LLIII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.73</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2623620960</td>\n",
       "      <td>MIT9211</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>LLIII</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>99.73</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>2681812902</td>\n",
       "      <td>MIT0918</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "      <td>LLII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.46</td>\n",
       "      <td>99.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>2623620962</td>\n",
       "      <td>MIT9303</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>2606217562</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>LLIV</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.46</td>\n",
       "      <td>99.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>2681813574</td>\n",
       "      <td>MIT1341</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00</td>\n",
       "      <td>99.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>2681813570</td>\n",
       "      <td>MIT1300</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.86</td>\n",
       "      <td>99.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>2681813572</td>\n",
       "      <td>MIT1307</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>LLVII</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>99.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>2681813568</td>\n",
       "      <td>MIT1223</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>LLVIII</td>\n",
       "      <td>LLI</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>99.18</td>\n",
       "      <td>99.74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID GenomeName     Type  CladeCyCOG CladeHackl  NContigsCyCOG  \\\n",
       "118  2623620345    MIT9515  ISOLATE         HLI        HLI              1   \n",
       "120  2606217259       MED4  ISOLATE         HLI        HLI              1   \n",
       "329  2681813573    MIT1314  ISOLATE        HLII       HLII              1   \n",
       "332  2606217688    MIT0604  ISOLATE        HLII       HLII              1   \n",
       "333  2606217559    MIT9215  ISOLATE        HLII       HLII              1   \n",
       "335  2606217708    MIT9312  ISOLATE        HLII       HLII              1   \n",
       "340  2623620959     AS9601  ISOLATE        HLII       HLII              1   \n",
       "343  2623620961    MIT9301  ISOLATE        HLII       HLII              1   \n",
       "470  2623620348     NATL1A  ISOLATE         LLI        LLI              1   \n",
       "471  2606217240     NATL2A  ISOLATE         LLI        LLI              1   \n",
       "472  2606217560    MIT0801  ISOLATE         LLI        LLI              1   \n",
       "473  2681812899    MIT0912  ISOLATE         LLI        LLI              1   \n",
       "474  2681812900    MIT0913  ISOLATE         LLI        LLI              1   \n",
       "475  2681812904    MIT1013  ISOLATE         LLI        LLI              1   \n",
       "476  2681813567    MIT1214  ISOLATE         LLI        LLI              1   \n",
       "478  2681812859    MIT0917  ISOLATE         LLI        LLI              1   \n",
       "479  2681812901    MIT0915  ISOLATE         LLI        LLI              1   \n",
       "510  2623620733      SS120  ISOLATE  LLII.LLIII       LLII              1   \n",
       "514  2681812903    MIT0919  ISOLATE  LLII.LLIII      LLIII              1   \n",
       "515  2623620960    MIT9211  ISOLATE  LLII.LLIII      LLIII              1   \n",
       "516  2681812902    MIT0918  ISOLATE  LLII.LLIII       LLII              1   \n",
       "536  2623620962    MIT9303  ISOLATE        LLIV       LLIV              1   \n",
       "543  2606217562    MIT9313  ISOLATE        LLIV       LLIV              1   \n",
       "599  2681813574    MIT1341  ISOLATE       LLVII      LLVII              1   \n",
       "600  2681813570    MIT1300  ISOLATE       LLVII      LLVII              1   \n",
       "601  2681813572    MIT1307  ISOLATE       LLVII      LLVII              1   \n",
       "603  2681813568    MIT1223  ISOLATE      LLVIII        LLI              1   \n",
       "\n",
       "     NContigsHackl  CompletenessCyCOG  CompletenessHackl  \n",
       "118              1             100.00             100.00  \n",
       "120              1              99.46             100.00  \n",
       "329              1             100.00             100.00  \n",
       "332              1              99.73              99.85  \n",
       "333              1              99.73             100.00  \n",
       "335              1              99.73             100.00  \n",
       "340              1              99.64              99.89  \n",
       "343              1              99.46             100.00  \n",
       "470              1              99.73             100.00  \n",
       "471              1              99.45             100.00  \n",
       "472              1              99.18              99.56  \n",
       "473              1              99.18              99.56  \n",
       "474              1              99.18             100.00  \n",
       "475              1              99.18             100.00  \n",
       "476              1              99.18             100.00  \n",
       "478              1              98.91             100.00  \n",
       "479              1              98.64              99.82  \n",
       "510              1              99.86             100.00  \n",
       "514              1              99.73             100.00  \n",
       "515              2              99.73             100.00  \n",
       "516              1              99.46              99.43  \n",
       "536              1             100.00             100.00  \n",
       "543              1              99.46              99.93  \n",
       "599              1             100.00              99.85  \n",
       "600              1              99.86              99.45  \n",
       "601              1              99.18              99.07  \n",
       "603              1              99.18              99.74  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at list of complete genomes with only one contig\n",
    "\n",
    "single_contig_genomes_df = overlap_df[(overlap_df['NContigsCyCOG'] == 1) | (overlap_df['NContigsHackl'] == 1)]\n",
    "print(f'There are {len(single_contig_genomes_df)} complete isolate genomes that are closed (one contig only) ' + \n",
    "      f'representing {single_contig_genomes_df.CladeHackl.nunique()} distinct clades.')\n",
    "\n",
    "single_contig_genomes_df[\n",
    "    ['GenomeID', 'GenomeName', 'Type', 'CladeCyCOG', 'CladeHackl', 'NContigsCyCOG', 'NContigsHackl', \n",
    "     'CompletenessCyCOG', 'CompletenessHackl']\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "99e6df05-7a46-4728-98dd-9bc1b3602b99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save genome-metadata.csv\n",
    "\n",
    "if not os.path.isdir('metadata'):\n",
    "    os.mkdir('metadata')\n",
    "genome_df.to_csv('metadata/genome-metadata.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b3aa5ec-8707-470f-9397-8177b6d834c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0c479b5-842b-495c-8e35-32e4f5d4c764",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d140a7f3-0049-4b4a-9b2c-6f3f6acf38c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3eee4d-d08e-47a0-8891-528ca947cc83",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
